barretenberg/cpp/src/barretenberg/stdlib/primitives/group/cycle_group.cpp

```diff
@@ -0,0 +1,1322 @@
+                                              can_unconditional_add);
+        offset_accumulator += offset_generator_delta;
+        if (has_fixed_points) {
+            result = can_unconditional_add ? result.unconditional_add(variable_accumulator)
+                                           : result.constrained_unconditional_add(variable_accumulator);
+        } else {
+            result = variable_accumulator;
+        }
+    }
+
+    // Update `result` to remove the offset generator terms, and add in any constant terms from `constant_acc`.
+    // We have two potential modes here:
+    // 1. All inputs are fixed-base and we constant_acc is not the point at infinity
+    // 2. Everything else.
+    // Case 1 is a special case, as we *know* we cannot hit incomplete addition edge cases,
+    // under the assumption that all input points are linearly independent of one another.
+    // Because constant_acc is not the point at infnity we know that at least 1 input scalar was not zero,
+    // i.e. the output will not be the point at infinity. We also know under case 1, we won't trigger the
+    // doubling formula either, as every point is lienarly independent of every other point (including offset
+    // generators).
+    if (!constant_acc.is_point_at_infinity() && can_unconditional_add) {
+        result = result.unconditional_add(AffineElement(-offset_accumulator));
+    } else {
+        // For case 2, we must use a full subtraction operation that handles all possible edge cases, as the output
+        // point may be the point at infinity.
+        // TODO(@zac-williamson) We can probably optimise this a bit actually. We might hit the point at infinity,
+        // but an honest prover won't trigger the doubling edge case.
+        // (doubling edge case implies input points are also the offset generator points,
+        // which we can assume an honest Prover will not do if we make this case produce unsatisfiable constraints)
+        // We could do the following:
+        // 1. If x-coords match, assert y-coords do not match
+        // 2. If x-coords match, return point at infinity, else return result - offset_accumulator.
+        // This would be slightly cheaper than operator- as we do not have to evaluate the double edge case.
+        result = result - AffineElement(offset_accumulator);
+    }
+    return result;
+}
+
+template <typename Composer> cycle_group<Composer> cycle_group<Composer>::operator*(const cycle_scalar& scalar) const
+{
+    return batch_mul({ scalar }, { *this });
+}
+
+template <typename Composer> cycle_group<Composer>& cycle_group<Composer>::operator*=(const cycle_scalar& scalar)
+{
+    *this = operator*(scalar);
+    return *this;
+}
+
+template <typename Composer>
+cycle_group<Composer> cycle_group<Composer>::operator/(const cycle_scalar& /*unused*/) const
+{
+    throw_or_abort("Implementation under construction...");
```

Comment:
TODO

Reply:
I added a TODO and assigned you to it. Hope you don't mind.